internal/provider/resources/resource_api_token.go

```diff
@@ -0,0 +1,339 @@
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/astronomer/terraform-provider-astro/internal/clients"
+	"github.com/astronomer/terraform-provider-astro/internal/clients/iam"
+	"github.com/astronomer/terraform-provider-astro/internal/provider/models"
+	"github.com/astronomer/terraform-provider-astro/internal/provider/schemas"
+	"github.com/astronomer/terraform-provider-astro/internal/utils"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/samber/lo"
+)
+
+// Ensure provider defined types fully satisfy framework interfaces.
+var _ resource.Resource = &ApiTokenResource{}
+var _ resource.ResourceWithImportState = &ApiTokenResource{}
+var _ resource.ResourceWithConfigure = &ApiTokenResource{}
+
+func NewApiTokenResource() resource.Resource {
+	return &ApiTokenResource{}
+}
+
+// ApiTokenResource defines the resource implementation.
+type ApiTokenResource struct {
+	IamClient      *iam.ClientWithResponses
+	OrganizationId string
+}
+
+func (r *ApiTokenResource) Metadata(
+	ctx context.Context,
+	req resource.MetadataRequest,
+	resp *resource.MetadataResponse,
+) {
+	resp.TypeName = req.ProviderTypeName + "_api_token"
+}
+
+func (r *ApiTokenResource) Schema(
+	ctx context.Context,
+	req resource.SchemaRequest,
+	resp *resource.SchemaResponse,
+) {
+	resp.Schema = schema.Schema{
+		// This description is used by the documentation generator and the language server.
+		MarkdownDescription: "API Token resource",
+		Attributes:          schemas.ApiTokenResourceSchemaAttributes(),
+	}
+}
+
+func (r *ApiTokenResource) Configure(
+	ctx context.Context,
+	req resource.ConfigureRequest,
+	resp *resource.ConfigureResponse,
+) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	apiClients, ok := req.ProviderData.(models.ApiClientsModel)
+	if !ok {
+		utils.ResourceApiClientConfigureError(ctx, req, resp)
+		return
+	}
+
+	r.IamClient = apiClients.IamClient
+	r.OrganizationId = apiClients.OrganizationId
+}
+
+func (r *ApiTokenResource) Create(
+	ctx context.Context,
+	req resource.CreateRequest,
+	resp *resource.CreateResponse,
+) {
+	var data models.ApiTokenResource
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var diags diag.Diagnostics
+
+	// Convert Terraform set of roles to API token roles
+	roles, diags := RequestApiTokenRoles(ctx, data.Roles)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	// Create the API token request
+	createApiTokenRequest := iam.CreateApiTokenRequest{
+		Name: data.Name.ValueString(),
+		Role: roles[0].Role,
```

Comment:
no guarantee that this will be the correct role
we should filter for the role depending on the token type and entity type of roles